[IMP] website: introduce `s_tabs_images` snippet

Brieuc-brd opened the pull request:
### Shares `s_tabs` options
This PR shares the `s_tabs` options so that they can be reused by snippets in the same category.

### Standardises `s_tabs` options and styles
This PR standardises the options and styles of `s_tabs` snippet so that they can be reused in other tabs snippets.
It also finetunes the style and options:
- Reviews the "focus-visible" shadow to make it correctly visible in all tab styles [*1]
- Reviews the text wrapping of tab items to ensure correct scrolling when the tab overflows the container [*2]
- Reviews the arrows of the Slide Effect buttons to match the correct animation.
- Maintains style consistency by replacing arbitrary values with variables.
- Removes unnecessary commented code.

### Introduces `s_tabs_images` snippet
This PR adds the new `s_tabs_images` snippet.

task-4094407
Part of task-4077427

Requires:
- https://github.com/odoo/design-themes/pull/1001
- https://github.com/odoo/upgrade/pull/6691

---

<img width="1372" alt="Capture d’écran 2024-10-29 à 10 04 38" src="https://github.com/user-attachments/assets/1a1066ca-1326-4b69-8bf0-7102c549181f">

[*1]
| Before | After |
|--------|--------|
| ![Capture d’écran 2025-02-14 à 16 30 09](https://github.com/user-attachments/assets/54264c58-1ba8-45d8-b414-968c7638e08b) | ![Capture d’écran 2025-02-14 à 16 30 34](https://github.com/user-attachments/assets/d0bde553-0a1f-40cc-bbcd-4b5c31042f0b) |

[*2]
| Before | After |
|--------|--------|
| ![Capture d’écran 2025-02-14 à 16 29 25](https://github.com/user-attachments/assets/c31fa59f-c2c5-43a7-a11b-56955a19d19b) | ![Capture d’écran 2025-02-14 à 16 28 11](https://github.com/user-attachments/assets/439cf242-76fe-4572-b2cb-ac3b99fb6333) |





---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/185642.png)](https://mergebot.odoo.com/odoo/odoo/pull/185642)

Brieuc-brd commented:
Hello @stefanorigano 👋 
This snippet is ready to be reviewed, could you take a look at it ? Thanks in advance !

lebl-odoo commented:
@Brieuc-brd 

Once we dropped the snippet, I encounter flickering when I switch for the first time between tabs. I suppose it's not directly linked to the snippet design, but we should address the issue (cc @odoo/rd-website ).

Regarding how to add images, we should stick to one solution.

Here is how we do for s_carousel
![image](https://github.com/user-attachments/assets/74ddc3df-2f41-4e7d-9fc4-51db243aebb1)

Here is how we do for this snippet
![image](https://github.com/user-attachments/assets/63aa5ee0-6e2b-43bf-844c-7626053e6d1b)

We should uniformize by adding a new container called "tab" with some options including those ones to add new elements. What do you think?

Brieuc-brd commented:
Hello @lebl-odoo 👋 
> Once we dropped the snippet, I encounter flickering when I switch for the first time between tabs. I suppose it's not directly linked to the snippet design, but we should address the issue (cc @odoo/rd-website ).

I haven't been able to reproduce it, but indeed this is something for the Website team I think. 🙂 

> We should uniformize by adding a new container called "tab" with some options including those ones to add new elements. What do you think?

I see your point and it make sense 👍.  But I see a problem, carousels and tabs are a bit different: the slides in the carousel take up the entire area of the carousel, which means that when you click on a carousel, you will always target the slide. Which is not the case for tabs snippets, the content of a tab doesn’t take the entire space.
So, if we move the Add/Remove buttons over the tab content, the users could miss them if they don't click in the right area.

| Clicking on the tab content | Clicking on the tab |
|--------|--------|
| <img width="1781" alt="Capture d’écran 2024-11-06 à 10 32 32" src="https://github.com/user-attachments/assets/a643b2f4-ec4c-4fab-9c1c-7a21a418e57a"> | <img width="1792" alt="Capture d’écran 2024-11-06 à 10 32 40" src="https://github.com/user-attachments/assets/d93f4915-daff-4efa-bba5-6b6dd7feb638"> |

Maybe that's why we originally placed them directly on the page snippet?

I'm not sure there's a solution to this issue, other than to review the whole structure of these snippets, or trigger these buttons in different places depending on which part of the snippet you clicked? 🤔 

What do you think @stefanorigano ?

In the meantime, I've made a small change to get the same button style than carousels 👍 :

| Before | After |
|--------|--------|
| <img width="283" alt="Capture d’écran 2024-11-06 à 11 01 31" src="https://github.com/user-attachments/assets/dd1a259f-be25-4c41-8353-3929e0bb09cc"> | <img width="289" alt="Capture d’écran 2024-11-06 à 11 02 24" src="https://github.com/user-attachments/assets/3d9c3d75-5e0f-4f42-b4aa-fa3c244d42f7"> |

lebl-odoo commented:
> But I see a problem, carousels and tabs are a bit different: the slides in the carousel take up the entire area of the carousel, which means that when you click on a carousel, you will always target the slide. Which is not the case for tabs snippets, the content of a tab doesn’t take the entire space.

What I've in mind is wherever I click, I got both "Tabs Images" and "Tab" section. However, if it is the only option we have in this section, it doesn't make sense.

Let's wait SRI's answer :)

Brieuc-brd commented:
@stefanorigano 👋 
Just a little reminder ☝️ 🙂

Brieuc-brd commented:
@stefanorigano 👋 
Small reminder 🙏

stefanorigano commented:
> Let's wait SRI's answer :)

Sorry for the late reply. In terms of UI, IMO we should remove the [+]/[-] button-group and introduce a "Tabs [Add New]" button.
The removal can be handled at "tab level", by clicking on the existing trash icon.

![image](https://github.com/user-attachments/assets/0617dddb-8b2b-4329-8cf9-d39507d5d994)

Sidenotes: I've noticed several usability issues with the current `s_tabs` snippet, and these issues are inherited by the snippet this PR aims to introduce. On top of those, this snippets comes with other issues my its own, like the edition of the tab's "description/subtitle" that's quite challenging and fragile. 

https://github.com/user-attachments/assets/af472ac1-0807-4f97-bf57-ab5f2b0074d4


@Brieuc-brd , I remember we've discussed these points at the office, but these issues are still reproducible on my side... while my review it's still requested. Does this need more work before reviewing ? Should we ask @lebl-odoo to open a task to fix `s_tabs` before jumping on this?

Brieuc-brd commented:
> I remember we've discussed these points at the office, but these issues are still reproducible on my side... while my review it's still requested. Does this need more work before reviewing ? Should we ask @lebl-odoo to open a task to fix s_tabs before jumping on this?

There's already an ongoing task for this issue 👍  : 
https://www.odoo.com/odoo/project/974/tasks/4316648

stefanorigano commented:
> There's already an ongoing task for this issue 👍 :

@Brieuc-brd , great, add me back as reviewer when ready then :)

stefanorigano commented on addons/website/static/src/snippets/s_tabs/003.scss:
```diff
@@ -0,0 +1,97 @@
+        }
+
+        .nav-underline {
+            border-bottom: 1px solid $border-color;
+
+            .nav-link {
+                margin-bottom: -1px;
```
It's unlikely that the default border thickness will ever change.. but who knows, maybe one day we'll link it to an option that the user can customize globally (?)... also, good practice and code readability ;)

```suggestion
            border-bottom: $border-width solid $border-color;

            .nav-link {
                margin-bottom: $border-width * -1;
```

stefanorigano commented on addons/website/static/src/snippets/s_tabs/003.scss:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    .s_tabs_main {
+        --card-cap-bg: initial;
```
why 🤔

anso-odoo replied:
Had a look and it looks safe to remove ([this line](https://github.com/odoo/odoo/commit/80fde992370e4474c27a383a61814ce1159f550c#diff-ecb18a1c6c2b57ad28b91104a9d9a392b10605fec61b3a9aa2c219c4c6eb4bf9R37) doing the job)

stefanorigano commented on addons/website/static/src/snippets/s_tabs_images/000.scss:
```diff
@@ -0,0 +1,10 @@
+.s_tabs_images {
+    .o_nav_tabs_description {
+        display: none;
+    }
```
Missing empty line between selectors

stefanorigano commented on addons/website/static/src/snippets/s_tabs_images/000.scss:
```diff
@@ -0,0 +1,10 @@
+    .o_nav_tabs_description {
+        display: none;
+    }
+    .s_tabs_main.s_tabs_nav_with_decriptions.s_tabs_nav_vertical .nav-underline .o_nav_tabs_description {
```
This selector looks unnecessarily strong... Could you try something like this?
```suggestion
    .s_tabs_nav_with_decriptions:where(.s_tabs_nav_vertical) :where(.nav-underline) where(.o_nav_tabs_description) {
```

stefanorigano reviewed (changes requested):
Good job @Brieuc-brd 🙌🙂
A lot of glitches here and there, but it's mostly fine on my my side.

Thanks so much 💪

Brieuc-brd commented on addons/website/views/snippets/s_tabs_images.xml:
```diff
@@ -0,0 +1,121 @@
+            <div class="s_tabs_main s_tabs_nav_vertical s_tabs_nav_with_decriptions s_col_no_resize s_col_no_bgcolor row">
+                <div class="s_tabs_nav col-sm-3 mb-3" data-name="Tab Header">
+                    <ul class="nav nav-underline flex-nowrap flex-sm-column overflow-y-hidden overflow-x-auto" role="tablist">
+                        <li class="nav-item" role="presentation">
```
This attribute has been added on `s_tabs` to improve the accessibility on website pages in : https://github.com/odoo/odoo/pull/156710/commits/92f345bdeab3cda29be83c9704dcf0b0543c9afd#diff-8d81e08e5ef6950f011929379cca8e4caa567bbc2e652195ac9252f3ce96e892R10
Since this snippet is based on `s_tabs` I kept the basic structure.

Do you think we should remove it 🤔

Brieuc-brd commented on addons/website/views/snippets/s_tabs_images.xml:
```diff
@@ -0,0 +1,121 @@
+                        <div class="oe_structure oe_empty">
+                            <section class="s_text_block" data-snippet="s_text_block" data-name="Tab">
+                                <div class="container s_allow_columns">
+                                    <img src="/web/image/website.s_tabs_images_default_image_5" class="img img-fluid rounded" alt="" style="width: 100% !important;"/>
```
I can't reproduce it. Do you remember the steps you followed to achieve it ? 🤔

Brieuc-brd replied:
@stefanorigano small reminder about your comment :) ☝️

lebl-odoo commented:
Hello @Brieuc-brd ,

Thanks for the nice snippet and the good job :muscle: 

I've some remarks:

1/ When I use the horizontal layout, description is hidden. When I press <delete> (and not backspace, sorry), I'm able to merge the hidden description with my title.

![2025-02-18_16-54](https://github.com/user-attachments/assets/6c8a4734-d42c-4e77-a109-bafd638154d9)

2/ Not sure how we should handle this, but some options can be irrelevant depending how the user uses it
![2025-02-18_16-54_1](https://github.com/user-attachments/assets/4f4c91e9-2d60-4dc7-b946-9cf6453f92c0)

Brieuc-brd commented on addons/website/views/snippets/s_tabs_images.xml:
```diff
@@ -0,0 +1,121 @@
+        <attribute name="data-select-class">s_tabs_nav_vertical</attribute>
+    </xpath>
+    <xpath expr="(//we-select)[2]" position="after">
+        <we-checkbox string="Descriptions" class="o_we_sublevel_1" data-dependencies="vertical_opt" data-select-class="s_tabs_nav_with_decriptions"/>
```
In the end, I added the possibility of having a description for the Pills style as well, what do you think ?
<img width="1446" alt="Capture d’écran 2025-02-21 à 10 24 39" src="https://github.com/user-attachments/assets/7601c031-37ea-4c9f-8f16-835cb20aff7d" />

Brieuc-brd commented:
Thanks for your review @stefanorigano.

Since some issues comes from `s_tabs` which has an inheritance on this new snippet, I’ve fixed these issues in the same commit which standardises the style and options. I’ve updated the commit message and the PR message.

Could you have a look when you have time ?

Here’s the diff : https://github.com/odoo/odoo/compare/041030d43fa46efd2365dc3061b083bbd445b1e8..0513ac9f455321a2bd4dbfe50853628d0ac0d597

Thanks in advance !


@lebl-odoo 👋 

> I found weird how the "Slide" option is working. I feel it's the exact oposite of what I'm expecting. In other words, if I select "bottom", it feels like I selected "top". Same thoughts about left/right.

Indeed you’re right 👍 I’ve update the buttons

> What do you think to put some ellipsis on text (both title and subtitle) to avoid those kind of weird stuffs users could do?

I’m not sure about that. As these elements are customisable (the user can add and remove different types of content using “/”), it’s up to the user. And for that reason, I’m not sure it’s possible to force the truncation.

To do something like that, I think we would have to add specific placeholders that the user couldn’t remove and force the truncation on them. It seems quite technical and will reduce possibilities of customisation for the user.


> Is there a reason why we couldn't set the "Button" in a vertical layout? For tabs, it make sense.

At first, the description was only for Underline style, but I’ve made some tests and it can work with Pills as well. I’ve added the possibility to add the description on Pills style.

> 1/ When I use the horizontal layout, description is hidden. When I press (and not backspace, sorry), I'm able to merge the hidden description with my title.
2/ Not sure how we should handle this, but some options can be irrelevant depending how the user uses it

I'm not sure how to handle that. We will take a look at this. Thanks for the review 👍

Brieuc-brd commented:
Hey @lebl-odoo 👋 

> 1/ When I use the horizontal layout, description is hidden. When I press (and not backspace, sorry), I'm able to merge the hidden description with my title.

Thanks to @bvr-odoo, this should be fixed 😄 .

> 2/ Not sure how we should handle this, but some options can be irrelevant depending how the user uses it

It seems that this is already the case for `s_tabs`, I don't know how to manage this either 🤔.
Having said that, is it common to have a scrolling navigation before selecting a style ? Most of the time, the user tests the options before integrating the content, if I'm not mistaken ... I'm not sure 🤔 .
But I agree that we should probably review this for all tab snippets in the future 👍 .

lebl-odoo commented:
Wonderful, I'll take the point. Let's move forward, thanks for the good work :clap:

SergeBayet commented:
> Hey @odoo/rd-website 👋 This snippet is ready for tech. review, could you have a look at it ? Thanks in advance !

@Brieuc-brd I'm on it. :+1:

SergeBayet commented:
@Brieuc-brd Could you rebase your branch before I review? Thx

SergeBayet commented on addons/website/static/src/js/editor/shared_options/tabs.js:
```diff
@@ -168,10 +161,10 @@ options.registry.NavTabsStyle = options.Class.extend({
         navLinks.forEach(link => link.classList.toggle('py-2', isVertical));
-        tabsNav.classList.toggle('col-sm-3', isVertical);
-        tabsContent.classList.toggle('col-sm-9', isVertical);
+        tabsNav.classList.toggle('col-md-3', isVertical);
+        tabsContent.classList.toggle('col-md-9', isVertical);
```
double quotes :)

Brieuc-brd commented:
Thanks for the review @SergeBayet 👍 
I've updated the branch.

SergeBayet commented on addons/website/views/snippets/s_tabs_images.xml:
```diff
@@ -0,0 +1,121 @@
+    <xpath expr="//we-button[@data-set-direction='vertical']" position="attributes">
+        <attribute name="data-select-class">s_tabs_nav_vertical</attribute>
+    </xpath>
+    <xpath expr="(//we-select)[2]" position="after">
+        <we-checkbox string="Descriptions" class="o_we_sublevel_1" data-select-class="s_tabs_nav_with_descriptions" data-name="description_opt" data-dependencies="vertical_opt"/>
+    </xpath>
```
It's odd to me to see a sub-option "Description" for the option "Direction". Don't you think?
![image](https://github.com/user-attachments/assets/a850fadb-db38-450e-a9ea-4f833f191fa5)

Edit: Ok "description" depends on "vertical direction"... My bad, but it still odd :)

SergeBayet replied:
Maybe remove the `o_we_sublevel_1`?
![image](https://github.com/user-attachments/assets/7d225b44-9dc2-4872-b937-a6eb1e3520ff)

SergeBayet reviewed (commented):
Looks good. Some remarks though.

SergeBayet reviewed (approved):
LGTM! :)

bso-odoo commented:
@robodoo delegate=SergeBayet

SergeBayet commented:
@robodoo r+

robodoo commented:
@Brieuc-brd @SergeBayet because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

SergeBayet commented:
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.
